src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3179,19 +3179,21 @@ protected DynamicObject stringAppend(DynamicObject string, DynamicObject other)
     public static abstract class StringAwkSplitPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
         @Child private RopeNodes.BytesNode bytesNode = RopeNodes.BytesNode.create();
+        @Child private YieldNode yieldNode = YieldNode.create();
         @Child RopeNodes.CodeRangeNode codeRangeNode = RopeNodes.CodeRangeNode.create();
         @Child private RopeNodes.GetCodePointNode getCodePointNode = RopeNodes.GetCodePointNode.create();
         @Child private SubstringNode substringNode = SubstringNode.create();
 
         private static final int SUBSTRING_CREATED = -1;
 
         @Specialization(guards = "is7Bit(string, codeRangeNode)")
-        protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit,
+        protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit, Object block,
```

Comment:
I think the implementations are different enough that this could be split into multiple specializations:
`protected DynamicObject stringAwkSplitSingleByteNoBlock(DynamicObject string, int limit, Nil block, ...`
`protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit, DynamicObject block, ...`
Thoughts @eregon ?

Reply:
I thought about it, but doesn't it require 4 methods in the end (because of guards)?

```java
@Specialization(guards = "is7Bit(string, codeRangeNode)")
protected methodWithBlock ...

@Specialization(guards = "!is7Bit(string, codeRangeNode)")
protected methodWithBlock ... 

@Specialization(guards = "is7Bit(string, codeRangeNode)")
protected methodNoBlock ... 

@Specialization(guards = "!is7Bit(string, codeRangeNode)")
protected methodNoBlock ...
```

Reply:
There is quite a bit of logic in the two specializations, so duplicating the bodies seem unfortunate.
How about profiling whether a block is passed with a ConditionProfile instead?
That way only one the two branches of addSubstring() will be compiled when always used the same way (always with block, or always without block, taking splitting into account).

Reply:
Done: https://github.com/oracle/truffleruby/pull/2052/commits/7c73068d1f4f9f37f3c767e3ef0f3766fd3ff26e